# This Python file uses the following encoding: utf-8
import sys, os
from PyQt6.QtWidgets import QApplication, QMainWindow, QMessageBox
from DownloadSingleVideo import Ui_DownloadSingleVideo
import pyperclip
import yt_dlp
from PyQt6.QtCore import QThread, pyqtSignal, QUrl, QTimer, Qt, QAbstractTableModel
from PyQt6.QtMultimedia import QMediaPlayer, QAudioOutput
from PyQt6.QtMultimediaWidgets import QVideoWidget
from DB import *
from sqlmodel import select, func
import requests
import pandas as pd

def save_comments_to_db(video_id: int, comments_list: list):
    session = Session(engine)

    for c in comments_list:
        comment = Comment(
            comid=int(c["id"]) if c["id"].isdigit() else 0,
            author=c["author"],
            text=c["text"],
            likecount=c["likecount"],
            video_id=video_id
        )
        session.add(comment)
    session.commit()
    print(f"âœ… ØªÙ… Ø­ÙØ¸ {len(comments_list)} ØªØ¹Ù„ÙŠÙ‚ ÙÙŠ Ù‚Ø§Ø¹Ø¯Ø© Ø§Ù„Ø¨ÙŠØ§Ù†Ø§Øª.")

def get_comments(video_url: str):
    ydl_opts = {
        'quiet': True,
        'skip_download': True,       # Ù„Ø§ ÙŠØ­Ù…Ù„ Ø§Ù„ÙÙŠØ¯ÙŠÙˆ Ù†ÙØ³Ù‡
        'getcomments': True,         # âœ… Ø¶Ø±ÙˆØ±ÙŠ Ù„ØªØ­Ù…ÙŠÙ„ Ø§Ù„ØªØ¹Ù„ÙŠÙ‚Ø§Øª
        'extract_flat': False,
    }

    with yt_dlp.YoutubeDL(ydl_opts) as ydl:
        info = ydl.extract_info(video_url, download=False)
    
    comments_data = []
    if 'comments' in info:
        for c in info['comments']:
            comments_data.append({
                "id": c.get("id"),
                "author": c.get("author", "Unknown"),
                "text": c.get("text", ""),
                "likecount": c.get("like_count", 0)
            })
    else:
        print("âš ï¸ Ù„Ø§ ØªÙˆØ¬Ø¯ ØªØ¹Ù„ÙŠÙ‚Ø§Øª ÙÙŠ Ù‡Ø°Ø§ Ø§Ù„ÙÙŠØ¯ÙŠÙˆ")

    return comments_data

class PandasModel(QAbstractTableModel):
    def __init__(self, df: pd.DataFrame, parent=None):
        super().__init__(parent)
        self._df = df

    def rowCount(self, parent=None):
        return len(self._df)

    def columnCount(self, parent=None):
        return len(self._df.columns)

    def data(self, index, role=Qt.ItemDataRole.DisplayRole):
        if index.isValid():
            if role == Qt.ItemDataRole.DisplayRole:
                value = self._df.iloc[index.row(), index.column()]
                return str(value)
        return None

    def headerData(self, section, orientation, role=Qt.ItemDataRole.DisplayRole):
        if role == Qt.ItemDataRole.DisplayRole:
            if orientation == Qt.Orientation.Horizontal:
                return str(self._df.columns[section])
            elif orientation == Qt.Orientation.Vertical:
                return str(self._df.index[section])
        return None


def download_image_as_bytes(url: str) -> bytes:
    if not url:
        return b""
    try:
        r = requests.get(url, timeout=10)
        r.raise_for_status()
        return r.content
    except Exception as e:
        print(f"âŒ ÙØ´Ù„ ØªØ­Ù…ÙŠÙ„ Ø§Ù„ØµÙˆØ±Ø©: {e}")
        return b""

def create_onevideo_from_d(d, thumb, playlist_id=None):
    # Ù†Ø­ØµÙ„ Ø¹Ù„Ù‰ Ø§Ù„ØµÙˆØ±Ø© ÙƒÙ€ bytes
    img_data = download_image_as_bytes(d['thumbnail'])

    # Ù†Ù†Ø´Ø¦ Ø§Ù„ÙƒØ§Ø¦Ù†
    video = OneVideo(
        videoid=d['id'],
        title=d['title'],
        desc=d['description'],
        filepath=thumb['filename'],
        img=img_data,
        playlist_id=playlist_id
    )
    return video

def LoadAllSingleVideos():
    if os.path.exists("downloads_temp/videos/"):
        return os.listdir("downloads_temp/videos/")
    else:
        return []

class DownloadThread(QThread):
    progress_changed = pyqtSignal(int)   # signal ÙŠØ±Ø³Ù„ Ø§Ù„Ù†Ø³Ø¨Ø© (0-100)
    quality: str = "1080p"

    def __init__(self, url, quality):
        super().__init__()
        self.url = url
        self.quality = quality
        self.ydl_opts = {
            # âœ… Ø¥Ø¹Ø¯Ø§Ø¯Ø§Øª Ø§Ù„ØªØ­Ù…ÙŠÙ„ Ø§Ù„Ø¹Ø§Ù…Ø©
            'format': 'bestvideo+bestaudio/best',
            'merge_output_format': 'mp4',
            'outtmpl': 'videos/%(title)s.%(ext)s',
            'noplaylist': False,
            'ignoreerrors': True,           # ØªØ¬Ø§Ù‡Ù„ Ø§Ù„ÙÙŠØ¯ÙŠÙˆÙ‡Ø§Øª Ø§Ù„Ù…Ø¹Ø·ÙˆØ¨Ø©
            'continuedl': True,             # Ø§Ø³ØªØ¦Ù†Ø§Ù Ø§Ù„ØªØ­Ù…ÙŠÙ„ Ù„Ùˆ ÙƒØ§Ù† Ù†Ø§Ù‚Øµ
            'retries': 10,                  # Ø¹Ø¯Ø¯ Ø§Ù„Ù…Ø­Ø§ÙˆÙ„Ø§Øª Ø¹Ù†Ø¯ Ø§Ù„ÙØ´Ù„
            'fragment_retries': 10,         # Ù…Ø­Ø§ÙˆÙ„Ø§Øª Ù„ÙƒÙ„ fragment
            'skip_unavailable_fragments': True,
            'keep_fragments': True,         # ÙŠØ­ØªÙØ¸ Ø¨Ø§Ù„Ù…Ù„ÙØ§Øª Ø§Ù„Ù…Ø¤Ù‚ØªØ© Ù„ÙŠØ³ØªØ£Ù†Ù Ù…Ù†Ù‡Ø§
            'file_access_retries': 5,       # Ø¥Ø¹Ø§Ø¯Ø© Ù…Ø­Ø§ÙˆÙ„Ø© Ø§Ù„ÙƒØªØ§Ø¨Ø© Ø¹Ù„Ù‰ Ø§Ù„Ù‚Ø±Øµ
            'noprogress': False,
            'concurrent_fragment_downloads': 5,  # ØªØ­Ù…ÙŠÙ„ Ø£Ø¬Ø²Ø§Ø¡ Ù…ØªØ¹Ø¯Ø¯Ø© Ù„ØªØ³Ø±ÙŠØ¹ Ø§Ù„Ø³Ø±Ø¹Ø©
            'socket_timeout': 30,                # ÙŠÙ…Ù†Ø¹ timeout Ø§Ù„Ø³Ø±ÙŠØ¹
            'writethumbnail': False,
            'writeinfojson': True,               # ÙŠØ­ÙØ¸ JSON Ø¨ÙƒÙ„ Ø¨ÙŠØ§Ù†Ø§Øª Ø§Ù„ÙÙŠØ¯ÙŠÙˆ
            'quiet': False,
            'verbose': False,
            # âœ… Ù…Ø¬Ù„Ø¯ Ù…Ø¤Ù‚Øª Ù„Ù„ØªØ­Ù…ÙŠÙ„ (ØªÙØ³ØªØ£Ù†Ù Ø§Ù„Ù…Ù„ÙØ§Øª Ù…Ù†Ù‡)
            'paths': {'home': os.path.join(os.getcwd(), "downloads_temp")},
        }

        if self.quality == "144p":
            self.ydl_opts['format'] = '160+140/best'
        elif self.quality == "360p":
            self.ydl_opts['format'] = '18/best'
        elif self.quality == "480p":
            self.ydl_opts['format'] = '135+140/best'
        elif self.quality == "720p":
            self.ydl_opts['format'] = '22/best'
        elif self.quality == "1080p":
            self.ydl_opts['format'] = '137+140/best'
        elif self.quality == "4K":
            self.ydl_opts['format'] = '313+251/best'
        elif self.quality == "ØµÙˆØª ÙÙ‚Ø· (m4a)":
            self.ydl_opts['format'] = '140'
        elif self.quality == "ØµÙˆØª ÙÙ‚Ø· (webm)":
            self.ydl_opts['format'] = '251'
        else:
            self.ydl_opts['format'] = 'bestvideo+bestaudio/best'

    def run(self):
        def progress_hook(d):
            if d['status'] == 'downloading':
                total = d.get('total_bytes') or d.get('total_bytes_estimate')
                downloaded = d.get('downloaded_bytes', 0)
                if total:
                    percent = int(downloaded * 100 / total)
                    self.progress_changed.emit(percent)
            elif d['status'] == 'finished':
                
                with yt_dlp.YoutubeDL(self.ydl_opts) as ydl:
                    info = ydl.extract_info(self.url, download=False)
                onevideo = create_onevideo_from_d(info, d)
                session.add(onevideo)
                session.commit()

                save_comments_to_db(info['id'], get_comments(self.url))

                
                self.progress_changed.emit(100)

        # Ø¥Ø¹Ø¯Ø§Ø¯ Ø®ÙŠØ§Ø±Ø§Øª yt_dlp
        self.ydl_opts['progress_hooks'] = [progress_hook]
        self.ydl_opts['continuedl'] = True
        self.ydl_opts['ignoreerrors'] = True
        self.ydl_opts['retries'] = 10
        self.ydl_opts['fragment_retries'] = 10

        try:
            with yt_dlp.YoutubeDL(self.ydl_opts) as ydl:
                ydl.download([self.url])
        except Exception as e:
            pass

class DownloadSingleVideo(QMainWindow):
    ThreadOf: DownloadThread

    def __init__(self, parent=None):
        super().__init__(parent)
        self.ui = Ui_DownloadSingleVideo()
        self.ui.setupUi(self)
        self.ui.downloadbtn.clicked.connect(self.DownloadVideo)
        self.ui.pastebtn.clicked.connect(self.PastLink)
        self.ThreadOf = None
        self.ui.Quality.addItems([
            "144p", "360p", "480p", "720p", "1080p", "4K",
            "Voice Only (m4a)", "Voice Only (webm)"
        ])
        self.ui.Quality.setCurrentText("1080p")

        
        # Ø¥Ù†Ø´Ø§Ø¡ Ø§Ù„Ù…Ø¤Ù‚Øª
        self.timer = QTimer(self)
        self.timer.setInterval(1000)  # ÙƒÙ„ 1000ms = Ø«Ø§Ù†ÙŠØ© ÙˆØ§Ø­Ø¯Ø©

        # Ø±Ø¨Ø· Ø§Ù„Ù…Ø¤Ù‚Øª Ø¨Ø¯Ø§Ù„Ø© (signal-slot)
        self.timer.timeout.connect(self.update_paths)

        # ØªØ´ØºÙŠÙ„ Ø§Ù„Ù…Ø¤Ù‚Øª
        self.timer.start()

        self.timer2 = QTimer(self)
        self.timer2.timeout.connect(self.update_progress)
        self.timer2.start(500)

        self.ui.play_button.clicked.connect(self.toggle_play)

        self.player = None

        self.is_playing = False

        # â¯ï¸ ØªØ´ØºÙŠÙ„ / Ø¥ÙŠÙ‚Ø§Ù Ù…Ø¤Ù‚Øª
    def toggle_play(self):
        if not self.player:
            return

        if self.is_playing:
            self.player.pause()
            self.ui.play_button.setText("â–¶ï¸")
        else:
            self.player.play()
            self.ui.play_button.setText("â¸ï¸")
        self.is_playing = not self.is_playing

    # ğŸ”Š ØªØºÙŠÙŠØ± Ø§Ù„ØµÙˆØª
    def change_volume(self, value):
        if not self.player:
            return
        
        self.audio_output.setVolume(value / 100)

    # ğŸ“ˆ ØªØ­Ø¯ÙŠØ« Ø´Ø±ÙŠØ· Ø§Ù„ØªÙ‚Ø¯Ù… ÙˆØ§Ù„ÙˆÙ‚Øª
    def update_progress(self):
        if not self.player:
            return
        duration = self.player.duration()
        position = self.player.position()
        if duration > 0:
            # progress = int((position / duration) * 1000)

            self.ui.time_label.setText(f"{self.format_time(position)} / {self.format_time(duration)}")

    # â© Ø¹Ù†Ø¯ ØªØ­Ø±ÙŠÙƒ Ø´Ø±ÙŠØ· Ø§Ù„ØªÙ‚Ø¯Ù…
    def seek(self, value):
        if not self.player:
            return
        
        duration = self.player.duration()
        if duration > 0:
            self.player.setPosition(int(duration * (value / 1000)))

    # â±ï¸ ØªØ­ÙˆÙŠÙ„ Ø§Ù„ÙˆÙ‚Øª Ø¥Ù„Ù‰ ØµÙŠØºØ© mm:ss
    def format_time(self, ms):
        seconds = ms // 1000
        minutes = seconds // 60
        seconds = seconds % 60
        return f"{minutes:02}:{seconds:02}"

    
    def selected_changed(self, selected, deselected):
        # Ù„Ùˆ Ù…Ø§ÙÙŠØ´ ØµÙÙˆÙ Ù…Ø­Ø¯Ø¯Ø©ØŒ Ù†Ø®Ø±Ø¬
        if not selected.indexes():
            return

        # Ù†Ø£Ø®Ø° Ø£ÙˆÙ„ ØµÙ Ù…Ø­Ø¯Ø¯ Ù…Ù† Ø§Ù„Ù…Ø¤Ø´Ø± Ø§Ù„Ù‚Ø§Ø¯Ù… Ù…Ù† Ø§Ù„Ø¥Ø´Ø§Ø±Ø©
        row = selected.indexes()[0].row()

        # Ù†ÙØªØ±Ø¶ Ø£Ù† Ø§Ù„Ù…ÙˆØ¯ÙŠÙ„ Ù…Ø±Ø¨ÙˆØ· Ø¨Ù€ self.ui.singlevideoshowtable
        model = self.ui.singlevideoshowtable.model()

        # Ù†Ø¨Ø­Ø« Ø¹Ù† Ø¹Ù…ÙˆØ¯ filepath (Ø£Ùˆ Ø§Ù„Ø¹Ù…ÙˆØ¯ Ø§Ù„Ø°ÙŠ ÙŠØ­ÙˆÙŠ Ø§Ù„Ù…Ø³Ø§Ø±)
        filepath_column = model.headerData(0, Qt.Orientation.Horizontal)
        filepath_col_index = None
        for i in range(model.columnCount()):
            if model.headerData(i, Qt.Orientation.Horizontal) == "filepath":
                filepath_col_index = i
                break

        if filepath_col_index is None:
            print("âš ï¸ Ù„Ù… ÙŠØªÙ… Ø§Ù„Ø¹Ø«ÙˆØ± Ø¹Ù„Ù‰ Ø¹Ù…ÙˆØ¯ filepath")
            return

        # Ù†Ø£Ø®Ø° Ù‚ÙŠÙ…Ø© Ø§Ù„Ù…Ø³Ø§Ø± Ù…Ù† Ø§Ù„Ù…ÙˆØ¯ÙŠÙ„
        video_path = model.index(row, filepath_col_index).data()

        self.ui.desc.setText(model.index(row, 3).data())
        self.ui.title.setText(model.index(row, 2).data())

        videoid = model.index(row, 1).data()

        self.ui.comments.setModel(PandasModel(pd.read_sql(f"Select author,text,likecount From comment Where video_id = '{videoid}'", engine)))

        self.ui.comments.resizeColumnsToContents()

        if not video_path:
            print("âš ï¸ Ù„Ø§ ÙŠÙˆØ¬Ø¯ Ù…Ø³Ø§Ø± ÙÙŠØ¯ÙŠÙˆ ÙÙŠ Ø§Ù„ØµÙ Ø§Ù„Ù…Ø­Ø¯Ø¯")
            return

        # ØªÙ‡ÙŠØ¦Ø© Ø§Ù„Ù…Ø´ØºÙ„ Ù…Ø±Ø© ÙˆØ§Ø­Ø¯Ø© ÙÙ‚Ø·
        if not self.player:
            self.player = QMediaPlayer(self)
            self.audio_output = QAudioOutput(self)
            self.player.setAudioOutput(self.audio_output)

            self.video_widget = QVideoWidget(self)
            self.player.setVideoOutput(self.video_widget)
            self.ui.playerlayout.addWidget(self.video_widget)

        # ØªØ´ØºÙŠÙ„ Ø§Ù„ÙÙŠØ¯ÙŠÙˆ
        print(f"ğŸ¬ ØªØ´ØºÙŠÙ„ Ø§Ù„ÙÙŠØ¯ÙŠÙˆ: {video_path}")
        self.player.setSource(QUrl.fromLocalFile(video_path))
        self.player.play()

    def update_paths(self):
        countof = session.exec(select(func.count()).select_from(OneVideo)).one()

        if not self.ui.singlevideoshowtable.model():
            self.ui.singlevideoshowtable.setModel(PandasModel(pd.read_sql("Select * From onevideo", engine)))
            self.ui.singlevideoshowtable.selectionModel().selectionChanged.connect(self.selected_changed)

        if countof != self.ui.singlevideoshowtable.model().rowCount():
            self.ui.singlevideoshowtable.setModel(PandasModel(pd.read_sql("Select * From onevideo", engine)))

    def DownloadVideo(self):
        url = self.ui.videourl.text()
        if 'youtube' in url:
            try:
                self.ThreadOf = DownloadThread(url, self.ui.Quality.currentText())
                self.thread = self.ThreadOf
                self.thread.progress_changed.connect(self.ui.ProgressBar.setValue)
                self.thread.start()
            except Exception as ex:
                QMessageBox.information(self, 'Error', f'Insert a valid youtube url: {ex}')  
        else:
            QMessageBox.information(self, 'Error', 'Insert a valid youtube url')

    def PastLink(self):
        pasted = pyperclip.paste()

        if 'youtube' in pasted:
            self.ui.videourl.setText(pyperclip.paste())

if __name__ == "__main__":
    app = QApplication([])
    window = DownloadSingleVideo()
    window.show()
    sys.exit(app.exec())
